Implement --check flag for gleam format --stdin

QuinnWilton opened the pull request:
The first commit also addresses some permission errors I was getting while trying to format files on Mac OS.

Here's what using the command looks like: https://gist.github.com/QuinnWilton/5fcb8b5a1ce8829dd851427f6cd150ff

I tried to keep my changes in `format::command` to a minimum, but if you'd like me to do some refactoring to DRY up some of the shared logic between the stdin and non-stdin case, let me know and I'm happy to :)

I also didn't see existing tests for the `format::command` anywhere, so I wasn't sure where to start with testing this. If you have suggestions I'm happy to give that a try.

lpil commented on src/error.rs:
```diff
@@ -1073,6 +1083,16 @@ but it cannot be found.",
+                input: FormatInput::STDIN,
+            } => {
+                let diagnostic = ProjectErrorDiagnostic {
+                    title: "STDIN has not been formatted".to_string(),
```
I had a look around and it seems the canonical name is stdin (lowercase), at least according to GNU, Microsoft and the C++ reference.

Could we make this and the constructor lowercase? Thanks

lpil commented on src/format/command.rs:
```diff
@@ -114,10 +118,20 @@ pub fn format_stdin(_check: bool) -> Result<(), Error> {
+    } else {
+        print!("{}", formatted);
+        Ok(())
+    }
```
Can we move this `if` to the `run` function, to match the logic for formatting files?

lpil commented on src/format/command.rs:
```diff
@@ -33,14 +33,18 @@ fn check_formatting(formatted_files: Vec<Formatted>) -> Result<(), Error> {
     for formatted in formatted_files {
         let path = formatted.path;
-        let mut f = File::open(&path).map_err(|e| Error::FileIO {
+        let mut f = File::create(&path).map_err(|e| Error::FileIO {
```
What does this change do?

QuinnWilton replied:
This opens the file with write permissions. Using `File::open` I was getting permission errors when it tried to write the newly formatted file, since `File::open` only requests read permissions.

lpil replied:
Thank you

bausano commented on src/error.rs:
```diff
@@ -1073,6 +1083,16 @@ but it cannot be found.",
+            } => {
+                let diagnostic = ProjectErrorDiagnostic {
+                    title: "STDIN has not been formatted".to_string(),
+                    label: "".to_string(),
```
Alternatively `String::new()`

bausano commented on src/format/command.rs:
```diff
@@ -33,14 +33,18 @@ fn check_formatting(formatted_files: Vec<Formatted>) -> Result<(), Error> {
     } else {
-        Err(Error::Format { problem_files })
+        Err(Error::Format {
+            input: FormatInput::Files {
```
Can be shortened to `input: FormatInput::Files { problem_files }`.

bausano commented on src/format/command.rs:
```diff
@@ -103,7 +107,7 @@ fn format_file(path: PathBuf) -> Result<Formatted, Error> {
 }
 
-pub fn format_stdin(_check: bool) -> Result<(), Error> {
+pub fn format_stdin(check: bool) -> Result<(), Error> {
```
What does the `check` argument refer to? Is there a more descriptive name?

lpil replied:
It refers to the `--check` flag. I've requests that this argument be removed from this function and moved to the higher level.

QuinnWilton commented:
I pushed up a new commit that addresses your comments. It's a bit of a bigger refactor, since I shifted the complexity of handling the stdin / file case out of the error handling layer, and into `format::command`, by introducing `source` and `destination` fields onto the `Formatted` struct.

This feels like a more natural place for this information to live, but I've left the previous version intact if you'd prefer I revert :)

I'll squash the last 3 commits together if everything looks good -- no sense having my rough work in the tree.

QuinnWilton commented on src/format/command.rs:
```diff
@@ -6,21 +6,29 @@ use std::str::FromStr;
 }
 
+#[derive(Debug, PartialEq)]
+pub enum FormatDestination {
```
Which is a more idiomatic name in Rust: `FormatDestination` or just `Destination`?

lpil replied:
No idea :) Do what you prefer!

QuinnWilton commented on src/format/command.rs:
```diff
@@ -97,13 +118,14 @@ fn format_file(path: PathBuf) -> Result<Formatted, Error> {
 }
 
-pub fn format_stdin(_check: bool) -> Result<(), Error> {
+pub fn read_and_format_stdin() -> Result<Vec<Formatted>, Error> {
```
Perhaps it'd be better to just return `Result<Formatted, Error>` and then wrap it in `vec!` at the callsite?

lpil replied:
I think that's a good idea :)

lpil commented on src/format/command.rs:
```diff
@@ -6,21 +6,29 @@ use std::str::FromStr;
+}
+
 pub fn run(stdin: bool, check: bool, files: Vec<String>) -> Result<(), Error> {
+    let formatted;
```
Can we assign directly to the variable please? I much prefer the style of using expressions.

```rust
let formatted = if stdin { // ....
```

QuinnWilton replied:
I didn't realize that was an option! I much prefer that too

lpil replied:
Honestly I didn't know you could not initialise a variable. I guess Rust can keep track of whether variables are initialised in order to make it safe. I still don't like it 😅

QuinnWilton replied:
Yeah, it bothered me too. I'm not used to C-style languages having nice flow-control constructs, so I just assumed it was the best I could do in Rust :P This language is growing on me a little bit.

lpil replied:
Rust is Haskell in C's clothing :)

lpil commented on src/format/command.rs:
```diff
@@ -6,21 +6,29 @@ use std::str::FromStr;
+    if check {
+        check_formatting(formatted)
+    } else {
+        write_formatted(formatted)
```
This is very nice! Thanks!

QuinnWilton commented:
Pushed up a squashed diff with the comments addressed!
